### ISSUE TITLE
Is there an API syntax node that allows developers to get error messages through error messages?

### ISSUE THREAD

**Issue by NMSAzulX** [dotnet/roslyn#58503]
**Brief description:**
Is there an API syntax node that allows developers to get error messages through error messages?

Describe your suggestion here.
I'm not sure if this is feasible. Hope to get the syntax node of the error according to the diagnostic information.

**Languages applicable:**   C#
Is your analyzer C# only? VB Only? Or Both?  C# Only

**Code example that the analyzer should report:**
```C#
var semantiModel = compilation.GetSemanticModel(tree);
var errors = semantiModel.GetDiagnostics();
foreach (var item in errors)
{
    var error = (CSDiagnostic)item;
    //suggestion
    var errorSyntaxNode = error .SyntaxNode;

}
```

**Comment by Youssef1313** [comment]
> Is there an API syntax node that allows developers to get error messages through error messages?

Do you mean getting the syntax node of a diagnostic?

- I see you have `var error = (CSDiagnostic)item;`, but `CSDiagnostic` isn't a public API. So are you doing something inside roslyn itself?
- The `Diagnostic` class has a `Location` property which has a `SourceSpan` property. You can pass this `SourceSpan` to `treeRoot.FindNode` so you get the syntax node.

**Comment by NMSAzulX** [comment]
@Youssef1313 Thanks for your method. I got the right node, and other information can be obtained through semantics.
